velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -153,6 +155,119 @@ TpchPlan TpchQueryBuilder::getQ1Plan() const {
   return context;
 }
 
+TpchPlan TpchQueryBuilder::getQ3Plan() const {
+  static const std::string kCustomer = "customer";
+  static const std::string kOrders = "orders";
+  static const std::string kLineitem = "lineitem";
+
+  std::vector<std::string> lineitemColumns = {
+      "l_shipdate", "l_orderkey", "l_extendedprice", "l_discount"};
+  std::vector<std::string> ordersColumns = {
+      "o_orderdate", "o_shippriority", "o_custkey", "o_orderkey"};
+  std::vector<std::string> customerColumns = {"c_custkey", "c_mktsegment"};
+
+  auto lineitemSelectedRowType = getRowType(kLineitem, lineitemColumns);
+  const auto& lineitemFileColumns = getFileColumnNames(kLineitem);
+  auto ordersSelectedRowType = getRowType(kOrders, ordersColumns);
+  const auto& ordersFileColumns = getFileColumnNames(kOrders);
+  auto customerSelectedRowType = getRowType(kCustomer, customerColumns);
+  const auto& customerFileColumns = getFileColumnNames(kCustomer);
+
+  // dateVal :2 = '1995-03-15'
+  const auto orderDate = "o_orderdate";
+  const auto shipDate = "l_shipdate";
+  std::string orderDateFilter;
+  std::string shipDateFilter;
+  std::string customerFilter;
+
+  // DWRF does not support Date type. Use Varchar instead.
+  if (ordersSelectedRowType->findChild(orderDate)->isVarchar()) {
+    orderDateFilter = "o_orderdate < '1995-03-15'";
+  } else {
+    orderDateFilter = "o_orderdate < '1995-03-15'::DATE";
+  }
+
+  if (lineitemSelectedRowType->findChild(shipDate)->isVarchar()) {
+    shipDateFilter = "l_shipdate > '1995-03-15'";
+  } else {
+    shipDateFilter = "l_shipdate > '1995-03-15'::DATE";
+  }
+
+  // MarketSegment filter value :1 = 'BUILDING'
+  customerFilter = "c_mktsegment = 'BUILDING'";
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId lineitemPlanNodeId;
+  core::PlanNodeId ordersPlanNodeId;
+  core::PlanNodeId customerPlanNodeId;
+
+  auto customers = PlanBuilder(planNodeIdGenerator)
+                       .tableScan(
+                           kCustomer,
+                           customerSelectedRowType,
+                           customerFileColumns,
+                           {customerFilter})
+                       .capturePlanNodeId(customerPlanNodeId)
+                       .planNode();
+
+  auto custkeyJoinNode =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(
+              kOrders,
+              ordersSelectedRowType,
+              ordersFileColumns,
+              {orderDateFilter})
+          .capturePlanNodeId(ordersPlanNodeId)
+          .hashJoin(
+              {"o_custkey"},
+              {"c_custkey"},
+              customers,
+              "",
+              {"o_orderdate", "o_shippriority", "o_orderkey"})
+          .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .localPartition(
```

Comment:
I think we can use the new API introduced here https://github.com/facebookincubator/velox/pull/1866